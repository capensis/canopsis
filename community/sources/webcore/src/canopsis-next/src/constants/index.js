export * from './recurrence-rule';
export * from './common';
export * from './entity';
export * from './alarm';
export * from './broadcast-message';
export * from './date';
export * from './dynamic-info';
export * from './event-filter';
export * from './pattern';
export * from './healthcheck';
export * from './idle-rule';
export * from './junit';
export * from './modal';
export * from './pbehavior';
export * from './permission';
export * from './remediation';
export * from './scenario';
export * from './snmp';
export * from './stats';
export * from './store';
export * from './widget';
export * from './kpi';
export * from './user';
export * from './service-weather';
export * from './map';
export * from './flowchart';
export * from './parameters';
export * from './declare-ticket';
export * from './link-rule';
export * from './external-data';
export * from './tag';
export * from './counter';
export * from './calendar';
export * from './icon';
export * from './availability';
