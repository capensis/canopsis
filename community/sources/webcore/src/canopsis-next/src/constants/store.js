export const ENTITIES_TYPES = {
  alarm: 'alarm',
  alarmDetails: 'alarmDetails',
  alarmDetailsSteps: 'alarmDetailsSteps',
  entity: 'entity',
  service: 'entityservices',
  weatherService: 'weatherService',
  pbehavior: 'pbehavior',
  event: 'event',
  userPreference: 'userPreference',
  group: 'group',
  view: 'view',
  viewTab: 'viewTab',
  widget: 'widget',
  widgetTemplate: 'widgetTemplate',
  stat: 'stat',
  user: 'user',
  role: 'role',
  eventFilter: 'eventFilter',
  metaAlarmRule: 'metaAlarmRule',
  snmpRule: 'snmpRule',
  dynamicInfo: 'dynamicInfo',
  broadcastMessage: 'broadcastMessage',
  playlist: 'playlist',
  pbehaviorExceptions: 'pbehaviorExceptions',
  pbehaviorTypes: 'pbehaviorTypes',
  pbehaviorReasons: 'pbehaviorReasons',
  remediationInstruction: 'remediationInstruction',
  remediationJob: 'remediationJob',
  remediationConfiguration: 'remediationConfiguration',
  remediationInstructionExecution: 'remediationInstructionExecution',
  remediationInstructionStats: 'remediationInstructionStats',
  scenario: 'scenario',
  entityCategory: 'entityCategory',
  testSuite: 'testSuite',
  testSuiteHistory: 'testSuiteHistory',
  stateSetting: 'stateSetting',
  idleRules: 'idleRules',
  flappingRules: 'flappingRules',
  resolveRules: 'resolveRules',
  filter: 'filter',
  ratingSettings: 'ratingSettings',
  pattern: 'pattern',
  map: 'map',
  alarmTag: 'alarmTag',
  shareToken: 'shareToken',
  declareTicketRule: 'declareTicketRule',
};
