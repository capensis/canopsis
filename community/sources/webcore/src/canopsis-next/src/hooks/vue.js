import { getCurrentInstance } from 'vue';

/**
 * Hook for get current component instance
 *
 * @return {Vue}
 */
export const useComponentInstance = () => {
  const instance = getCurrentInstance();

  return instance?.proxy || instance;
};

/**
 * Hook for get current component options
 *
 * @return {Object}
 */
export const useComponentOptions = () => {
  const instance = useComponentInstance();

  return instance.$options;
};

/**
 * Hook for get current component model
 *
 * @return {Object}
 */
export const useComponentModel = () => {
  const { model } = useComponentOptions();

  return {
    event: model?.event ?? 'input',
    prop: model?.event ?? 'value',
  };
};
