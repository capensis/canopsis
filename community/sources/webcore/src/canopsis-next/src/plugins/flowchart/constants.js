export const DIRECTIONS = {
  northWest: 'nw',
  north: 'n',
  northEast: 'ne',
  east: 'e',
  southEast: 'se',
  south: 's',
  southWest: 'sw',
  west: 'w',
};
