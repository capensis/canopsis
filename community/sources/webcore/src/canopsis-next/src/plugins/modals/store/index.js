import Vue from 'vue';

import { VUETIFY_ANIMATION_DELAY } from '@/config';

import uid from '@/helpers/uid';

export const types = {
  SHOW: 'SHOW',
  HIDE: 'HIDE',
  HIDE_COMPLETED: 'HIDE_COMPLETED',
  MINIMIZE: 'MINIMIZE',
  MAXIMIZE: 'MAXIMIZE',
};

export default {
  namespaced: true,
  state: {
    allIds: [],
    byId: {},
  },
  getters: {
    modals: state => state.allIds.map(id => state.byId[id]),
    hasMaximizedModal: state => state.allIds.some(id => !state.byId[id].minimized),
  },
  mutations: {
    [types.SHOW](state, {
      id,
      name,
      config = {},
      dialogProps = {},
    }) {
      Vue.set(state.byId, id, {
        id,
        name,
        config,
        dialogProps,
        hidden: false,
        minimized: false,
      });

      state.allIds.push(id);
    },
    [types.HIDE](state, { id }) {
      Vue.set(state.byId[id], 'hidden', true);
    },
    [types.HIDE_COMPLETED](state, { id }) {
      state.allIds = state.allIds.filter(value => value !== id);

      Vue.delete(state.byId, id);
    },
    [types.MINIMIZE](state, { id }) {
      Vue.set(state.byId[id], 'minimized', true);
    },
    [types.MAXIMIZE](state, { id }) {
      Vue.set(state.byId[id], 'minimized', false);
    },
  },
  actions: {
    /**
     * Show modal window by name with unique id
     *
     * @param {Function} commit
     * @param {Object} state
     * @param {string} name
     * @param {Object} [config = {}]
     * @param {Object} [dialogProps = {}]
     * @param {string} [id = uid()]
     */
    show({ commit, state }, {
      name,
      config = {},
      dialogProps = {},
      id = uid('modal'),
    } = {}) {
      if (state.byId[id]) {
        return commit(types.MAXIMIZE, { id });
      }

      return commit(types.SHOW, {
        id,
        name,
        config,
        dialogProps,
      });
    },

    /**
     * Hide modal by id
     *
     * @param {Function} commit
     * @param {Object} state
     * @param {string} [id]
     */
    hide({ commit, state }, { id } = {}) {
      if (!id) {
        throw new Error('Missed required parameter');
      }

      if (!state.byId[id]) {
        return;
      }

      commit(types.HIDE, { id });

      /**
       * This function added for vuetify animation waiting
       */
      setTimeout(() => {
        if (state.byId[id]) {
          commit(types.HIDE_COMPLETED, { id });
        }
      }, VUETIFY_ANIMATION_DELAY);
    },
    /**
     * Minimize modal by id
     *
     * @param {Function} commit
     * @param {string} [id]
     */
    minimize({ commit }, { id } = {}) {
      if (!id) {
        throw new Error('Missed required parameter');
      }

      commit(types.MINIMIZE, { id });
    },

    /**
     * Maximize modal by id
     *
     * @param {Function} commit
     * @param {string} [id]
     */
    maximize({ commit }, { id } = {}) {
      if (!id) {
        throw new Error('Missed required parameter');
      }

      commit(types.MAXIMIZE, { id });
    },
  },
};
