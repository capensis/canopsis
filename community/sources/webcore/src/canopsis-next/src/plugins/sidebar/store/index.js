import { VUETIFY_ANIMATION_DELAY } from '@/config';

export const types = {
  SHOW: 'SHOW',
  HIDE: 'HIDE',
  HIDE_COMPLETED: 'HIDE_COMPLETED',
};

export default {
  namespaced: true,
  state: {
    name: null,
    config: {},
    hidden: true,
  },
  getters: {
    sidebar: state => state,
  },
  mutations: {
    [types.SHOW](state, { name, config = {} }) {
      state.name = name;
      state.config = config;
      state.hidden = false;
    },
    [types.HIDE](state) {
      state.hidden = true;
    },
    [types.HIDE_COMPLETED](state) {
      state.name = null;
      state.config = {};
    },
  },
  actions: {
    show({ commit }, { name, config = {} } = {}) {
      commit(types.SHOW, { name, config });
    },

    hide({ commit, state }) {
      commit(types.HIDE);

      /**
       * This function added for vuetify animation waiting
       */
      setTimeout(() => {
        if (state.hidden) {
          commit(types.HIDE_COMPLETED);
        }
      }, VUETIFY_ANIMATION_DELAY);
    },
  },
};
