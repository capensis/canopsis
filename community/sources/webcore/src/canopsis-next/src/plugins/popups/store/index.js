import { POPUP_TYPES } from '@/constants';
import { POPUP_AUTO_CLOSE_DELAY } from '@/config';

import { uid } from '@/helpers/uid';

export const types = {
  ADD: 'ADD',
  REMOVE: 'REMOVE',
  SET_DEFAULT_CLOSE_TIME: 'SET_DEFAULT_CLOSE_TIME',
};

export default {
  namespaced: true,
  state: {
    defaultCloseTimeByType: {
      [POPUP_TYPES.error]: POPUP_AUTO_CLOSE_DELAY,
      [POPUP_TYPES.success]: POPUP_AUTO_CLOSE_DELAY,
      [POPUP_TYPES.info]: POPUP_AUTO_CLOSE_DELAY,
      [POPUP_TYPES.warning]: POPUP_AUTO_CLOSE_DELAY,
    },
    popups: [],
  },
  getters: {
    popups: state => state.popups,
  },
  mutations: {
    [types.ADD](state, { popup }) {
      state.popups.push(popup);
    },
    [types.REMOVE](state, { id }) {
      state.popups = state.popups.filter(v => v.id !== id);
    },
    [types.SET_DEFAULT_CLOSE_TIME](state, { type, time }) {
      state.defaultCloseTimeByType[type] = time;
    },
  },
  actions: {
    add({ commit, state }, {
      id = uid('popup'),
      type,
      text,
      autoClose = state.defaultCloseTimeByType[type],
    } = {}) {
      commit(types.ADD, {
        popup: {
          id, type, text, autoClose,
        },
      });
    },

    setDefaultCloseTime({ commit }, { type, time = POPUP_AUTO_CLOSE_DELAY }) {
      return commit(types.SET_DEFAULT_CLOSE_TIME, { type, time });
    },

    remove({ commit }, { id }) {
      commit(types.REMOVE, { id });
    },

    success({ dispatch }, popup) {
      return dispatch('add', { ...popup, type: POPUP_TYPES.success });
    },

    info({ dispatch }, popup) {
      return dispatch('add', { ...popup, type: POPUP_TYPES.info });
    },

    warning({ dispatch }, popup) {
      return dispatch('add', { ...popup, type: POPUP_TYPES.warning });
    },

    error({ dispatch }, popup) {
      return dispatch('add', { ...popup, type: POPUP_TYPES.error });
    },
  },
};
