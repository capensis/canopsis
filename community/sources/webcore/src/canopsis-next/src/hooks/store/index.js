import { useComponentInstance } from '../vue';

/**
 * Normalizes a given map into an array of objects with `key` and `value` properties.
 * If the input is an array, each element is treated as both the key and value.
 * If the input is an object, it converts the entries into an array of objects.
 *
 * @param {Object|Array} map - The map to normalize, which can be either an object or an array.
 * @returns {Array.<{key: string, value: string}>} An array of objects with `key` and `value` properties.
 */
const normalizeStoreMap = (map = {}) => (
  Array.isArray(map)
    ? map.map(key => ({ key, value: key }))
    : Object.entries(map).map(([key, value]) => ({ key, value }))
);

/**
 * Retrieves the Vuex store instance from the current Vue component instance.
 * This function must be used within the Vue component's `setup()` method.
 *
 * @returns {Object} The Vuex store instance.
 * @throws {Error} If the function is not used within the `setup()` method.
 *
 * @example
 * // In a Vue component
 * import { useStore } from `./path/to/useStore`;
 *
 * export default {
 *   setup() {
 *     const store = useStore();
 *     // Now you can use the store to commit mutations, dispatch actions, or access state
 *     return {
 *       // Your other setup properties
 *     };
 *   }
 * }
 */
export const useStore = () => {
  const vm = useComponentInstance();

  if (!vm) {
    throw new Error('You must use this function within the "setup()" method');
  }

  return vm.$store;
};

/**
 * Creates hooks for accessing Vuex store module's getters and actions.
 * This function helps in creating a more modular and organized way to interact with Vuex store modules.
 *
 * @param {string} namespace - The namespace of the Vuex store module.
 * @returns {Object} An object containing the store, module, and functions to access getters and actions.
 * @throws {Error} If the provided namespace does not correspond to any module in the store.
 *
 * @example
 * // Assuming a Vuex store module is registered under `user/` namespace
 * import { useStoreModuleHooks } from `./path/to/useStoreModuleHooks`;
 *
 * export default {
 *   setup() {
 *     const { useGetters, useActions } = useStoreModuleHooks('user/');
 *     const { userName } = useGetters(['userName']);
 *     const { updateUser } = useActions({ customUpdateUser: 'updateUser' });
 *
 *     return {
 *       userName,
 *       customUpdateUser
 *     };
 *   }
 * }
 */
export const useStoreModuleHooks = (namespace) => {
  const store = useStore();
  const preparedNamespace = namespace.slice(-1) === '/' ? namespace : `${namespace}/`;
  const module = store._modulesNamespaceMap[preparedNamespace];

  if (!module) {
    throw new Error(`Incorrect module namespace = ${namespace}`);
  }

  const useGetters = getters => normalizeStoreMap(getters).reduce((acc, { key, value }) => {
    acc[key] = () => store.getters(preparedNamespace + value);

    return acc;
  }, {});

  const useActions = actions => normalizeStoreMap(actions).reduce((acc, { key, value }) => {
    acc[key] = (...args) => module.context.dispatch.apply(store, [value, ...args]);

    return acc;
  }, {});

  return {
    store,
    module,
    useGetters,
    useActions,
  };
};
