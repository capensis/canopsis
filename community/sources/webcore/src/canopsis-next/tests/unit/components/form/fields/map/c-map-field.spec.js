import flushPromises from 'flush-promises';
import Faker from 'faker';

import { generateShallowRenderer, generateRenderer } from '@unit/utils/vue';
import { createMockedStoreModules } from '@unit/utils/store';
import { createSelectInputStub } from '@unit/stubs/input';
import { MAX_LIMIT } from '@/constants';

import CMapField from '@/components/forms/fields/map/c-map-field.vue';

const stubs = {
  'v-select': createSelectInputStub('v-select'),
};

const factory = generateShallowRenderer(CMapField, {

  stubs,
  attachTo: document.body,
});

const snapshotFactory = generateRenderer(CMapField, {

  attachTo: document.body,
});

const selectSelectField = wrapper => wrapper.find('select.v-select');

describe('c-map-field', () => {
  const fetchMapsListWithoutStore = jest.fn().mockReturnValue({
    meta: {},
    data: [],
  });
  const mapModule = {
    name: 'map',
    actions: {
      fetchListWithoutStore: fetchMapsListWithoutStore,
    },
  };
  const store = createMockedStoreModules([
    mapModule,
  ]);

  afterEach(() => {
    fetchMapsListWithoutStore.mockClear();
  });

  it('Filters fetched after mount', async () => {
    factory({
      store,
      propsData: {},
    });

    await flushPromises();

    expect(fetchMapsListWithoutStore).toBeCalledTimes(1);
    expect(fetchMapsListWithoutStore).toBeCalledWith(
      expect.any(Object),
      { params: { limit: MAX_LIMIT } },
      undefined,
    );
  });

  it('Map changed after trigger select field', () => {
    const map = { _id: Faker.datatype.string() };
    fetchMapsListWithoutStore.mockReturnValueOnce({
      data: [map],
    });

    const wrapper = factory({
      propsData: {
        value: '',
      },
      store: createMockedStoreModules([mapModule]),
    });

    const selectField = selectSelectField(wrapper);

    selectField.vm.$emit('input', map._id);

    expect(wrapper).toEmit('input', map._id);
  });

  it('Renders `c-map-field` with default props', () => {
    const wrapper = snapshotFactory({
      store: createMockedStoreModules([mapModule]),
    });

    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper).toMatchMenuSnapshot();
  });

  it('Renders `c-map-field` with custom props', async () => {
    fetchMapsListWithoutStore.mockReturnValueOnce({
      data: [
        { _id: 'map-id-1' },
        { _id: 'map-id-2' },
      ],
    });

    const wrapper = snapshotFactory({
      store: createMockedStoreModules([mapModule]),
      propsData: {
        value: 'map-id-1',
        label: 'Custom label',
        name: 'customName',
        disabled: true,
        required: true,
        hideDetails: true,
      },
    });

    await flushPromises();

    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper).toMatchMenuSnapshot();
  });
});
