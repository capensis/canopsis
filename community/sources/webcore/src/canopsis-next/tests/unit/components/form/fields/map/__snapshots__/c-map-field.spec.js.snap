// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-map-field Renders \`c-map-field\` with default custom props 1`] = `
<div
  class="v-input v-text-field v-select v-input--hide-details v-input--is-disabled theme--light"
  role="combobox"
>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-select__slot"
      >
        <label
          aria-hidden="true"
          class="v-label v-label--is-disabled theme--light"
          style="left: 0px; position: absolute;"
        >
          Custom label
        </label>
        <div
          class="v-select__selections"
        >
          <input
            aria-label="Custom label"
            aria-readonly="false"
            autocomplete="on"
            disabled="disabled"
            name="customName"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--"
          >
            <i
              aria-hidden="true"
              class="v-icon v-icon--disabled v-icon--link material-icons theme--light"
            />
          </div>
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon v-icon--disabled material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      />
    </div>
  </div>
</div>
`;

exports[`c-map-field Renders \`c-map-field\` with default custom props 2`] = `
<div
  class="v-menu__content theme--light "
  style="max-height: 300px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
>
  <div
    class="v-select-list v-card theme--light"
  >
    <div
      class="v-list theme--light"
      role="list"
    >
      <div
        role="listitem"
      >
        <div
          class="v-list__tile theme--light"
        >
          <div
            class="v-list__tile__content"
          >
            <div
              class="v-list__tile__title"
            >
              No data available
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`c-map-field Renders \`c-map-field\` with default props 1`] = `
<div
  class="v-input v-text-field v-select theme--light"
  role="combobox"
>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-select__slot"
      >
        <label
          aria-hidden="true"
          class="v-label theme--light"
          style="left: 0px; position: absolute;"
        >
          Map
        </label>
        <div
          class="v-select__selections"
        >
          <input
            aria-label="Map"
            aria-readonly="false"
            autocomplete="on"
            name="map"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--"
          >
            <i
              aria-hidden="true"
              class="v-icon v-icon--link material-icons theme--light"
            />
          </div>
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      />
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <div
          class="v-messages__wrapper"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`c-map-field Renders \`c-map-field\` with default props 2`] = `
<div
  class="v-menu__content theme--light "
  style="max-height: 300px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
>
  <div
    class="v-select-list v-card theme--light"
  >
    <div
      class="v-list theme--light"
      role="list"
    >
      <div
        role="listitem"
      >
        <div
          class="v-list__tile theme--light"
        >
          <div
            class="v-list__tile__content"
          >
            <div
              class="v-list__tile__title"
            >
              No data available
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
